homeassistant/components/binary_sensor/timed_state_infer.py

```diff
@@ -0,0 +1,142 @@
+"""Infers its state from the current state and duration of other sensors."""
+import asyncio
+import logging
+from datetime import timedelta
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.const import CONF_ENTITY_ID, STATE_UNKNOWN, CONF_NAME
+import homeassistant.helpers.config_validation as cv
+from homeassistant.core import callback
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
+from homeassistant.helpers.event import async_track_state_change,\
+    async_track_point_in_time
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_TIME_ON = 'seconds_on'
+CONF_TIME_OFF = 'seconds_off'
+CONF_VALUE_ON = 'value_on'
+CONF_VALUE_OFF = 'value_off'
+DEFAULT_NAME = "Timed State Infer Binary Sensor"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Required(CONF_TIME_ON): cv.positive_int,
+    vol.Required(CONF_TIME_OFF): cv.positive_int,
+    vol.Required(CONF_VALUE_ON): vol.Coerce(float),
+    vol.Required(CONF_VALUE_OFF): vol.Coerce(float)
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Timed State Infer Binary sensor."""
+    async_add_devices([TimedStateInferBinarySensor(hass, config[CONF_NAME],
+                                                   config[CONF_ENTITY_ID],
+                                                   config[CONF_TIME_ON],
+                                                   config[CONF_TIME_OFF],
+                                                   config[CONF_VALUE_ON],
+                                                   config[CONF_VALUE_OFF])])
+
+
+class TimedStateInferBinarySensor(BinarySensorDevice):
+    """Representation of a sensor."""
+
+    def __init__(self, hass, name, observed_entity_id, time_on, time_off,
+                 value_on, value_off):
+        """Initialization of the Binary Sensor."""
+        self._hass = hass
+        self._name = name
+        self._observed_entity_id = observed_entity_id
+        self._time_on = timedelta(seconds=time_on)
+        self._time_off = timedelta(seconds=time_off)
+        self._value_on = value_on
+        self._value_off = value_off
+        self._is_on = False
+        self._pending = False
+        self._pending_since = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if sensor is on."""
+        return self._is_on
+
+    @asyncio.coroutine
```

Comment:
We've migrated to `async`/`await` syntax. Please drop this decorator and prefix async methods with `async` like `async def async_setup_platform`. Replace any `yield from` with `await`.